batchflow/config.py

```diff
@@ -1,289 +1,245 @@
-            self.config = config.config
-        else:
-            raise TypeError(f'config must be dict, Config or list but {type(config)} was given')
-
+            super().__init__(config)
+        elif isinstance(config, (dict, list, tuple)):
+            self.parse(config)
+        
         for key, value in kwargs.items():
             self.put(key, value)
-
-    def pop(self, variables, config=None, **kwargs):
-        """ Returns variables and remove them from config
+   
+    def parse(self, config):
+        """ Parses flatten config with slashes
 
         Parameters
         ----------
-        variables : str or list of strs
-            names of variables. '/' is used to get value from nested dict
-        config : dict, Config or None
-            if None, variables will be getted from self.config else from config
+        config : dict, Config, list or tuple
 
         Returns
         -------
-        single value or a tuple
+        self : dict
         """
-        if isinstance(config, Config):
-            value = config.pop(variables, None, **kwargs)
+        if isinstance(config, dict):
+            items = config.items()
+        elif isinstance(config, (tuple, list)):
+            items = config
+            if any([not isinstance(item, (tuple, list)) for item in items]):
+                raise ValueError('tuple or list should contain only tuples or lists')
+            if any([len(item) != 2 for item in items]):
+                raise ValueError('tuples in list should represent pairs key-value'
+                                 ', and therefore must be always the length of 2')
+
+        for key, value in items:
+            if isinstance(key, str):
+                key = '/'.join(filter(None, key.split('/')))
+            self.put(key, value)
+
+        return self
+      
+    def put(self, key, value):
+        """ Put a new key into config
+
+        Parameters
+        ----------
+        key : hashable object
+            key to add. '/' is used to put value into nested dict
+        value : masc
+        """
+        if isinstance(key, str) and '/' in key:
+            config = self
+            parent, child = key.split('/', 1)
+
+            if parent in config and isinstance(config[parent], Config):
+                config[parent].update(Config({child: value}))
+            else:
+                config[parent] = Config({child: value})
+
         else:
-            value = self._get(variables, config, pop=True, **kwargs)
+            if key in self and isinstance(self[key], dict) and isinstance(value, dict):
+                self[key].update(Config(value))
+            else:
+                super().__setitem__(key, value)
+
+    def __getitem__(self, key):
+        value = self._get(key)
         return value
 
-    def get(self, variables, default=None, config=None):
-        """ Returns variables from config
+    def __setitem__(self, key, value):
+        if key in self:
+            self.pop(key)
```

Comment:
> Also, why we can't write the `put` method in the way when we needn't `pop` before `put`?

I think, we can't, because we want to update value if key is already in config. However, there are some cases when pop is necessary. 
For example: `config  = Config({'a/b': 1, 'a/c': 2})`
 When we want to set a new value for `key='a'`, `config['a'] = dict(b=0, d=3)`, we want to get
 `config = {'a': {'b': 0, 'd': 3}}`.